src/2d/shallow/flag2refine2.f90

```diff
@@ -177,16 +177,35 @@ subroutine flag2refine2(mx,my,mbc,mbuff,meqn,maux,xlower,ylower,dx,dy,t,level, &
                     eta = q(1,i,j) + aux(1,i,j)
 
                     ! Check wave criteria
-                    if (abs(eta - sea_level) > wave_tolerance) then
-                        ! Check to see if we are near shore
-                        if (q(1,i,j) < deep_depth) then
-                            amrflags(i,j) = DOFLAG
-                            cycle x_loop
-                        ! Check if we are allowed to flag in deep water
-                        ! anyway
-                        else if (level < max_level_deep) then
-                            amrflags(i,j) = DOFLAG
-                            cycle x_loop
+		    ! New method - level specific wave tolerance
+                    do m=1,min(size(wave_tolerance),mxnest)
+                        if (abs(eta - sea_level) > wave_tolerance(m) .and. level <= m) then
+                            ! Check to see if we are near shore
+                            if (q(1,i,j) < deep_depth) then
+                                amrflags(i,j) = DOFLAG
+                                cycle x_loop
+                            ! Check if we are allowed to flag in deep water
+                            ! anyway
+                            else if (level < max_level_deep) then
+                                amrflags(i,j) = DOFLAG
+                                cycle x_loop
+                            endif
+                        endif
+                    enddo
+		    ! Check if wave tolerance array size is smaller than max level
+		    ! then refine with the last wave tolerance value for the rest of the levels
+		    if (size(wave_tolerance) < mxnest-1) then
```

Comment:
I might suggest pre-processing this so that the `wave_tolerance` array is always at least `mxnest` so you can simplify this logic and not replicate the checks.